stixcore/tmtc/packets.py

```diff
@@ -0,0 +1,317 @@
+
+        self.source_packet_header = data
+
+
+class TMPacket(GenericPacket):
+    """
+
+    """
+    def __init__(self, data):
+        """
+
+        Parameters
+        ----------
+        data
+        """
+        super().__init__(data)
+        self.data_header = TMDataHeader(self.source_packet_header.bitstream)
+
+    @classmethod
+    def is_datasource_for(cls, sph):
+        return not (sph.process_id == 90 and sph.packet_category == 12)
+
+
+class TCPacket(GenericPacket):
+    """
+    A TC packet class can hold any TC packets
+    """
+    def __init__(self, data):
+        """
+        Create a TCPacket give the data
+        Parameters
+        ----------
+        data : binary or `stixcore.tmtc.packets.SourcePacketHeader`
+            Data to create TC packet from
+        """
+        if not isinstance(data, SourcePacketHeader):
+            data = SourcePacketHeader(data)
+
+        self.source_packet_header = data
+        self.data_header = TMDataHeader(self.source_packet_header.bitstream)
+
+    @classmethod
+    def is_datasource_for(cls, sph):
+        return sph.packet_category == 12 and sph.process_id == 90
+
+
+class GenericTMPacket:
+    """
+    A generic TM packet all specific TM packets are subclasses of this class
+
+    Attributes
+    ----------
+    _registry : `dict`
+        Dictionary mapping classes (key) to function (value) which validates input.
+
+    Parameters
+    ----------
+    data : binary data or `stixcore.tmtc.packets.TMPacket`
+        Data to create TM packet from
+
+    Notes
+    -----
+    """
+    _registry = dict()
+
+    def __init_subclass__(cls, **kwargs):
+        """
+        An __init_subclass__ hook initializes all of the subclasses of a given class.
+        So for each subclass, it will call this block of code on import.
+        This registers each subclass in a dict that has the `is_datasource_for` attribute.
+        This is then passed into the factory so we can register them.
+        """
+        super().__init_subclass__(**kwargs)
+        if hasattr(cls, 'is_datasource_for'):
+            cls._registry[cls] = cls.is_datasource_for
+
+    def __init__(self, data):
+        """
+
+        Parameters
+        ----------
+        data
+        """
+        if not isinstance(data, TMPacket):
+            self.source_packet_header = SourcePacketHeader(data)
+            self.data_header = TMDataHeader(self.source_packet_header.bitstream)
+        else:
+            self.source_packet_header = data.source_packet_header
+            self.data_header = data.data_header
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}({self.source_packet_header}, {self.data_header})'
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class TM_1_1(GenericTMPacket):
+    """
+    TM(1,1) Telecommand acceptance report
+    """
+    def __init__(self, data):
+        super().__init__(data)
+        structure = {'NIX00001': 'uint:16', 'NIX00002': 'uint:16'}
```

Comment:
@drhlxiao would your IDB routines be able to give back the packet structure based on service type, service subtype and SSID?